src/cargo/ops/cargo_rustc/layout.rs

```diff
@@ -85,14 +85,33 @@ impl Layout {
     /// Differs from `at` in that this calculates the root path from the workspace target directory,
     /// adding the target triple and the profile (debug, release, ...).
     pub fn new(ws: &Workspace, triple: Option<&str>, dest: &str) -> CargoResult<Layout> {
+        use std::hash::{Hash, Hasher};
+        use std::collections::hash_map::DefaultHasher;
+
         let mut path = ws.target_dir();
         // Flexible target specifications often point at filenames, so interpret
         // the target triple as a Path and then just use the file stem as the
         // component for the directory name.
         if let Some(triple) = triple {
-            path.push(Path::new(triple)
-                .file_stem()
-                .ok_or_else(|| format_err!("target was empty"))?);
+            if triple.ends_with(".json") {
+                let json_path = Path::new(triple).canonicalize().map_err(|err| {
+                    format_err!("invalid target file path {:?}: {:?}", triple, err)
+                })?;
+                let file_stem = json_path
+                    .file_stem()
+                    .ok_or_else(|| format_err!("target was empty"))?
+                    .to_str()
+                    .ok_or_else(|| format_err!("target path is not valid unicode"))?;
+                let mut hasher = DefaultHasher::new();
+                json_path.hash(&mut hasher);
+                let hash = hasher.finish();
+
+                path.push(Path::new(&format!("{}-{}", file_stem, hash)));
```

Comment:
Hm are we sure that this is what we want to do? If we did this I think we may run a risk of not being able to easily discover the output filename, right?

Reply:
Oh, good point. I'll push a new version that keeps the old output folder names.